pilot/cmd/istioctl/config.go

```diff
@@ -0,0 +1,158 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/spf13/cobra"
+	"istio.io/istio/pkg/log"
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/tools/remotecommand"
+)
+
+var (
+
+	// TODO - Add support for non-kube istio deployments
+	// TODO - Add config-diff to get the difference between what pilot wants and what the proxy has
+	// TODO - Add support for non-default proxy config locations
+	configCmd = &cobra.Command{
+		Use:   "config <podID>",
+		Short: "Retrieves proxy configuration for the specified pod [kube only]",
+		Long: `
+Retrieves the proxy configuration for the specified pod when running in Kubernetes.
+Support for other environments to follow.
+`,
+		Example: `  # Retrieve config for productpage-v1-bb8d5cbc7-k7qbm pod
+  istioctl config productpage-v1-bb8d5cbc7-k7qbm`,
+		Aliases: []string{"conf"},
+		Args:    cobra.MinimumNArgs(1),
+		RunE: func(c *cobra.Command, args []string) error {
+			podName := args[0]
+			log.Infof("Retrieving proxy config for %q", podName)
+
+			ns := namespace
+			if ns == "" {
+				ns = defaultNamespace
+			}
+			config, err := readConfigFile(podName, ns)
+			if err != nil {
+				return err
+			}
+
+			fmt.Println(config)
+
+			return nil
+		},
+	}
+)
+
+func init() {
+	rootCmd.AddCommand(configCmd)
+}
+
+func createCoreV1Client() (*rest.RESTClient, error) {
+	config, err := defaultRestConfig()
+	if err != nil {
+		return nil, err
+	}
+	return rest.RESTClientFor(config)
+}
+
+func defaultRestConfig() (*rest.Config, error) {
+	config, err := clientcmd.BuildConfigFromFlags("", kubeconfig)
+	if err != nil {
+		return nil, err
+	}
+	config.APIPath = "/api"
+	config.GroupVersion = &v1.SchemeGroupVersion
+	config.NegotiatedSerializer = serializer.DirectCodecFactory{CodecFactory: scheme.Codecs}
+	return config, nil
+}
+
+func readConfigFile(podName, podNamespace string) (string, error) {
+	// Get filename to read from
+	var lsOut, lsErr bytes.Buffer
+	cmd := []string{"ls", "-Art", "/etc/istio/proxy"}
+	err := podExec(podName, podNamespace, cmd, nil, &lsOut, &lsErr)
```

Comment:
Can you make `podExec` return `stdout` and `stderr` instead of passing a pointer? It would simplify the code a bit and let you reduce the scope of `stdout` and `stderr` and avoid the need for `lsOut/lsErr/catOut/catErr`.

```golang
if stdout, stderr, err:= podExec(podName, podNamespace, cmd); err != nil {
    return "", err
} else if stderr.String() != "" {
    // return error ...
}
```

Reply:
`stdout` is needed outside of this scope so it would have to be something like this?
```
var stdout io.Writer
if stdout, stderr, err:= podExec(podName, podNamespace, cmd); err != nil {
    return "", err
} else if stderr.String() != "" {
    // return error ...
}
// do something with stdout
```
I think this would result in variable shadowing?